support-workers/src/typescript/lambdas/createSalesforceContactLambda.ts

```diff
@@ -0,0 +1,154 @@
+import type { CreateZuoraSubscriptionState } from '../model/createZuoraSubscriptionState';
+import { stageFromEnvironment } from '../model/stage';
+import type {
+	CreateSalesforceContactState,
+	WrappedState,
+} from '../model/stateSchemas';
+import {
+	createSalesforceContactStateSchema,
+	wrapperSchemaForState,
+} from '../model/stateSchemas';
+import { ServiceProvider } from '../services/config';
+import type { SalesforceContactRecords } from '../services/salesforce';
+import { SalesforceService } from '../services/salesforce';
+import { getSalesforceConfig } from '../services/salesforceClient';
+import { user } from '../test/fixtures/salesforceFixtures';
+import { asRetryError } from '../util/errorHandler';
+import { getIfDefined } from '../util/nullAndUndefined';
+
+const stage = stageFromEnvironment();
+const salesforceServiceProvider = new ServiceProvider(stage, async (stage) => {
+	const config = await getSalesforceConfig(stage);
+	return new SalesforceService(config);
+});
+
+export const handler = async (
+	state: WrappedState<CreateSalesforceContactState>,
+) => {
+	try {
+		console.info(`Input is ${JSON.stringify(state)}`);
+		const createSalesforceContactState = wrapperSchemaForState(
+			createSalesforceContactStateSchema,
+		).parse(state).state;
+		const serviceForUser = await salesforceServiceProvider.getServiceForUser(
+			state.state.user.isTestUser,
+		);
+		const contactRecords = await serviceForUser.createContactRecords(
+			createSalesforceContactState.user,
+			createSalesforceContactState.giftRecipient,
+		);
+		return {
+			...state,
+			state: createNextState(createSalesforceContactState, contactRecords),
+		};
+	} catch (error) {
+		throw asRetryError(error);
+	}
+};
+
+const createNextState = (
+	state: CreateSalesforceContactState,
+	contactRecords: SalesforceContactRecords,
+): CreateZuoraSubscriptionState => {
+	switch (state.product.productType) {
+		case 'Contribution':
+			return {
+				productSpecificState: {
+					productType: 'Contribution',
+					product: state.product,
+					paymentMethod: state.paymentMethod,
+					salesForceContact: contactRecords.buyer,
+					similarProductsConsent: state.similarProductsConsent,
+				},
+				...state,
```

Comment:
There are a couple of pieces which are common to all branches of this switch - nesting the product specific state under `productSpecificState` and merging with `state` at the top level. Would it be worth building up the product specific state separately, and then doing the common pieces in one place which applies to all products?

Reply:
I had a go at doing this but ran into difficulties convincing TS that the product returned from the sub function was the same type as the one in the calling function. It didn't save much code anyway because we still need the specific productType as mentioned in a previous comment so I've left it as is.